src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,49 +1,130 @@
+        }
+        if (index < elementData.length) {
+            System.arraycopy(elementData, index, elementData,
+                            index + 1, arraySize - index);
+            elementData[index] = value;
+            arraySize++;
+        } else if (index >= elementData.length) {
+            T[] newElementData = grow();
+            System.arraycopy(elementData, 0, newElementData,
+                    0, index - 1);
+            newElementData[index] = value;
+            System.arraycopy(elementData, index, newElementData,
+                    index + 1, arraySize - index);
+            elementData[arraySize++] = (T) newElementData;
+        } else {
+            elementData[arraySize++] = value;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        T[] newElementData = (T[]) new Object[list.size()];
+        for (int i = 0; i < newElementData.length; i++) {
+            newElementData[i] = list.get(i);
+        }
+        if (newElementData.length + arraySize > elementData.length) {
+            newElementData = grow();
+            System.arraycopy(elementData, 0, newElementData,
+                            0, arraySize);
+            System.arraycopy(elementData, 0, newElementData,
+                            0, newElementData.length);
+            elementData = newElementData;
+        } else {
+            System.arraycopy(newElementData, 0, elementData,
+                             arraySize, newElementData.length);
+        }
+        arraySize += newElementData.length;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index < 0 || index > arraySize - 1) {
+            throw new ArrayListIndexOutOfBoundsException("Index is invalid");
+        }
+        return elementData[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (index < 0 || index > arraySize - 1) {
+            throw new ArrayListIndexOutOfBoundsException("Index is invalid");
+        }
+        elementData[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index < 0 || index > arraySize) {
+            throw new ArrayListIndexOutOfBoundsException("Index is invalid");
+        }
+        T elementToRemove = elementData[index];
+        System.arraycopy(elementData, index + 1, elementData, index,
+                        arraySize - index - 1);
+        arraySize--;
+        return elementToRemove;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        boolean isElementExist = false;
+        T elementToRemove = null;
+        for (int i = 0; i < arraySize; i++) {
+            if (elementData[i] == null && element == null) {
+                elementToRemove = remove(i);
+                return elementToRemove;
+            } else if (elementData[i] != null && element.equals(elementData[i])) {
```

Comment:
can we combine logic from if block with logic from else block?

Reply:
done 
using Objects.equals()